### DISCUSSION TITLE
Py2 vs Py3 server installations

### DISCUSSION THREAD

**Discussion by ernstpijper** [DIRACGrid/DIRAC#5923] [Q&A] [answered]
Hi,

I read here on github the following:

```
As of DIRAC 7.2 there is also experimental support for Python 3 based clients (which are to considered 
production level fromDIRAC 7.3). As of DIRAC 7.3 there is also experimental support for Python 3 based 
servers (which are to considered production level from DIRAC 8.0). We recommend transitioning to Python 3 
clients during using DIRAC 7.2 and Python 3 servers using DIRAC 7.3.
```

So support for python 3 servers in 7.3 is experimental but you *do* recommend transitioning to python 3 for 7.3 server installations. 

So for 7.3 server installations, what exactly do you recommend?

Ernst

**Comment by fstagni** [comment, marked as answer]
Hi Ernst,
I think that when we wrote the sentences above we were over-cautious. I think you can read the sentence as-is while just removing the word "experimental". 

In fact, we may say that:
- you can have fully working python3 clients as of later versions of DIRAC 7.2 (say, 7.2.30+)
- you can have fully working python3 servers as of later versions of DIRAC 7.3 (say, 7.3.10+)

LHCb moved to python3 servers while using DIRAC 7.3.x (about 2 months ago) and have recently fully removed the python2 option.

Let me know if you need more clarifications.

**Reply by ernstpijper** [reply]
Hi Federico,

Thanks for the clarification.

I'm trying to install 7.3.17 with python3, according to these instructions:

https://dirac.readthedocs.io/en/latest/AdministratorGuide/ServerInstallations/InstallingDiracServer.html#primary-server-installation

I'm running the command

```
./install_site.sh -i /opt/dirac -v 7.3.17 install.cfg
```

At the end I'm getting

```
installation finished.
/opt/dirac/versions/v7.3.17-1646832011/Linux-x86_64/diracosrc: line 108: MAMBA_ROOT_PREFIX: unbound variable
```

I attached the complete output. 
[install.txt](https://github.com/DIRACGrid/DIRAC/files/8215167/install.txt)

The install.cfg I'm using is pretty much the same as for the python2 installation, slightly adapted from looking at the differences between the python2 and python3 examples in the documentation.

What I don't see is the installation of the different systems, services, databases and agents. So, obviously I'm not doing it right.

What am I missing?

**Reply by fstagni** [reply]
Hi Enrst, you just ran into into this bug: https://github.com/DIRACGrid/DIRACOS2/issues/69. Wait some minutes and try again with a new version of DIRACOS2 being built.

**Reply by ernstpijper** [reply]
It still doesn't work. I'll try again tomorrow. I can see the changes that Chris made in create_diracosrc.sh but somehow it has no effect yet on my installation attempt.

**Reply by fstagni** [reply]
I have just tried (in a fresh container) and worked fine for me. Can you make sure that the diracos directory is removed before trying?

**Reply by ernstpijper** [reply]
It nows works. No idea what has changed. I always made sure that when I ran install_site.sh, I first deleted everything related to the dirac installation in the installation path, /opt/dirac in my case.

Anyway, ran into some other problems. I'm running the install_site.sh script from another directory than the installation path. This goes wrong when the script performs a grep on install.cfg. The grep is done in the installation path but install.cfg is located somewhere else. So had to change this line in the install_site.sh script:

```
install_cfg=${1}
```

into this line

```
install_cfg=${PWD}/${1}
```

Then it works again. But now I'm running into a mysql problem:

```
2022-03-10 09:41:33 UTC dirac-setup-site [140512030754624] ERROR: Cannot connect to the DB  MySQL Error ( 1131 : Could not connect: (2002, "Can't connect to local MySQL server through socket '/tmp/mysql.sock' (2)"))
2022-03-10 09:41:33 UTC dirac-setup-site [140512030754624] ERROR: Could not connect to MySQL server
ERROR: Could not connect to MySQL server
```

It seems mysql assumes the socket file is located under /tmp. In reality it is located in /var/lib/mysql/mysql.sock. I realise I could change the location to /tmp/mysql.sock but this doesn't seem like a robust solution. Looks like it is caused by dirac using its own mysql command located in /opt/dirac/pro/Linux-x86_64/bin/mysql which assumes a default location for the socket file.

What to do?

**Reply by fstagni** [reply]
I don't think you can have 2 separate installations in 2 different directories. CC @chrisburr

**Reply by ernstpijper** [reply]
I made a link from 

```
/tmp/mysql.sock -> /var/lib/mysql/mysql.sock
```

and now the install_site.sh runs succesfully. When it's done i see a list of all the services that are running.

I'm not trying to have 2 separate dirac installations in 2 different directories, if that's what you mean. The install_site.sh script offers the '-i' options, which is the path where you want to install dirac. This suggests that you can run the install_site.sh script from another directory then where the dirac installation takes place. Without the '-i' option, install_site.sh installs dirac in the directory from where it is run. 

However when doing this

```
cd /dir/a
install_site.sh -i /opt/dirac install.cfg
```
the script changes directory to /opt/dirac and then later does a grep for install.cfg:

```
if [[ "$(grep WebPortal "${install_cfg}" | cut -d '=' -f 2 | tr -d " ")" == "yes" ]]; then
  pip_install_args+=("WebAppDIRAC[server]")
fi
```
but in my case install.cfg is located in directory /dir/a, and so the grep fails. When I make change above, it works again.

**Reply by fstagni** [reply]
I let @chrisburr or @atsareg reply on this one. 

I think that for client installations the "install where you want" makes sense, but for servers I would stick to `/opt/dirac`, as there are other assumptions taking place.

**Reply by ernstpijper** [reply]
Hi Federico,

It seems I have a working py3 dirac server installed. Tested it with only 2 jobs so far but it looks good.

One thing that I noticed is that in version 7.3.19, the MinorStatus stays in 'Pending Requests' when the job is done. This is also true for 7.3.18. In 7.3.17, MinorStatus becomes 'Execution Complete' when the jobs is done.
